File: legacy/src/lib/ast/serializer.ts
import { ASTNode } from './types'

export function serialize(nodes: ASTNode[]): string {
  return nodes.map(serializeNode).join('')
}

function serializeNode(node: ASTNode): string {
  switch (node.type) {
    case 'number':   return node.value
    case 'operator': return node.op === '×' ? '*' : node.op === '÷' ? '/' : node.op
    case 'constant': return node.name === 'Ans' ? 'ans' : node.name
    case 'fraction': return `(${serialize(node.numerator)})/(${serialize(node.denominator)})`
    case 'exponent': return `(${serialize(node.base)})^(${serialize(node.exponent)})`
    case 'radical':
      return node.degree.length === 0
        ? `sqrt(${serialize(node.radicand)})`
        : `nthRoot(${serialize(node.radicand)}, ${serialize(node.degree)})`
    case 'function':
      if (node.name === 'log')   return `log(${serialize(node.argument)}, 10)`
      if (node.name === 'ln')    return `log(${serialize(node.argument)})`
      if (node.name === 'pow10') return `10^(${serialize(node.argument)})`
      if (node.name === 'exp')   return `e^(${serialize(node.argument)})`
      return `${node.name}(${serialize(node.argument)})`
    case 'factorial': return `factorial(${serialize(node.operand)})`
    case 'ncr':       return `combinations(${serialize(node.n)}, ${serialize(node.r)})`
    case 'npr':       return `permutations(${serialize(node.n)}, ${serialize(node.r)})`
    case 'paren':     return `(${serialize(node.children)})`
  }
}

File: lib/core/ast/serializer.dart
import 'types.dart';

String serialize(List<ASTNode> nodes) {
  return nodes.map(_serializeNode).join('');
}

String _serializeNode(ASTNode node) {
  switch (node) {
    case NumberNode n:
      return n.value;
    case OperatorNode n:
      if (n.op == '×') return '*';
      if (n.op == '÷') return '/';
      return n.op;
    case ConstantNode n:
      return n.name == 'Ans' ? 'ans' : n.name;
    case FractionNode n:
      return '(${serialize(n.numerator)})/(${serialize(n.denominator)})';
    case ExponentNode n:
      return '(${serialize(n.base)})^(${serialize(n.exponent)})';
    case RadicalNode n:
      if (n.degree.isEmpty) {
        return 'sqrt(${serialize(n.radicand)})';
      }
      return 'nthRoot(${serialize(n.radicand)}, ${serialize(n.degree)})';
    case FunctionNode n:
      if (n.name == FunctionName.log) {
        return 'log(${serialize(n.argument)}, 10)';
      }
      if (n.name == FunctionName.ln) {
        return 'log(${serialize(n.argument)})';
      }
      if (n.name == FunctionName.pow10) {
        return '10^(${serialize(n.argument)})';
      }
      if (n.name == FunctionName.exp) {
        return 'e^(${serialize(n.argument)})';
      }
      return '${n.name.name}(${serialize(n.argument)})';
    case FactorialNode n:
      return 'factorial(${serialize(n.operand)})';
    case NcrNode n:
      return 'combinations(${serialize(n.n)}, ${serialize(n.r)})';
    case NprNode n:
      return 'permutations(${serialize(n.n)}, ${serialize(n.r)})';
    case ParenNode n:
      return '(${serialize(n.children)})';
  }
}
